sed/config/config_model.py

```diff
@@ -0,0 +1,262 @@
+class BinningModel(BaseModel):
+    hist_mode: str
+    mode: str
+    pbar: bool
+    threads_per_worker: int
+    threadpool_API: str
+
+
+class HistogramModel(BaseModel):
+    bins: Sequence[int]
+    axes: Sequence[str]
+    ranges: Sequence[tuple[float, float]]
+
+
+class StaticModel(BaseModel):
+    """Static configuration settings that shouldn't be changed by users."""
+
+    # flash specific settings
+    stream_name_prefixes: Optional[dict] = None
+    stream_name_postfixes: Optional[dict] = None
+    beamtime_dir: Optional[dict] = None
+
+
+class EnergyModel(BaseModel):
+    bins: int
+    ranges: Sequence[int]
+    normalize: bool
+    normalize_span: int
+    normalize_order: int
+    fastdtw_radius: int
+    peak_window: int
+    calibration_method: str
+    energy_scale: str
+    tof_fermi: int
+    tof_width: Sequence[int]
+    x_width: Sequence[int]
+    y_width: Sequence[int]
+    color_clip: int
+
+    class EnergyCalibrationModel(BaseModel):
+        d: float
+        t0: float
+        E0: float
+        energy_scale: str
+
+    calibration: Optional[EnergyCalibrationModel] = None
+
+    class EnergyCorrectionModel(BaseModel):
+        correction_type: str
+        amplitude: float
+        center: Sequence[float]
+        gamma: float
+        sigma: float
+        diameter: float
+
+    correction: Optional[EnergyCorrectionModel] = None
+
+
+class MomentumModel(BaseModel):
+    axes: Sequence[str]
+    bins: Sequence[int]
+    ranges: Sequence[Sequence[int]]
+    detector_ranges: Sequence[Sequence[int]]
+    center_pixel: Sequence[int]
+    sigma: int
+    fwhm: int
+    sigma_radius: int
+
+    class MomentumCalibrationModel(BaseModel):
+        kx_scale: float
+        ky_scale: float
+        x_center: float
+        y_center: float
+        rstart: float
+        cstart: float
+        rstep: float
+        cstep: float
+
+    calibration: Optional[MomentumCalibrationModel] = None
+
+    class MomentumCorrectionModel(BaseModel):
+        feature_points: Sequence[Sequence[float]]
+        rotation_symmetry: int
+        include_center: bool
+        use_center: bool
+
+    correction: Optional[MomentumCorrectionModel] = None
+
+
+class DelayModel(BaseModel):
+    adc_range: Sequence[int]
+    flip_time_axis: bool
+    # Group keys in the datafile
+    p1_key: Optional[str] = None
+    p2_key: Optional[str] = None
+    p3_key: Optional[str] = None
+    t0_key: Optional[str] = None
+
+    class Calibration(BaseModel):
+        creation_date: datetime
+        adc_range: Sequence[int]
+        delay_range: Sequence[float]
+        time0: float
+        delay_range_mm: Sequence[float]
+        datafile: FilePath  # .h5 extension in filepath
+
+    calibration: Optional[Calibration] = None
+
+    class Offsets(BaseModel):
+        creation_date: Optional[datetime] = None
+        constant: Optional[float] = None
+        flip_delay_axis: Optional[bool] = False
+
+        ## This seems rather complicated way to define offsets,
```

Comment:
As mentioned here, offsets has quite a convoluted config structure, and it differs from the passable args. The add_offsets method thus mostly tries to figure out what is being passed. I'd suggest we try refactoring the method and fix a specific config model for it.
I'd probably suggest to remove the dict like column entries completely.

Reply:
I can have a look. In principle, the structure is not so complicated. There are top level entries like constant, flip etc., and then a sub dictionary for each column with options like weights, preserve mean etc.

Reply:
The parameters are a bit different than the config dict, that is true, but the are refined into that structure. I think your config model looks okay, not sure I understand the problem correctly.

Reply:
I just found the way the value unpacking rather convoluted, and was trying to refactor it. But naturally, that is not what this PR is about.